### PR TITLE
feat: implement new state caches

### PR DESCRIPTION
**Pull request by twoeths**
**Motivation**

As part of n-historical states strategy, we want to keep bounded states in memory instead of unbounded states like what we have for now, see #6008

**Description**
- Unify term for out state cache: BlockStateCache vs CheckpointStateCache
- BlockStateCache interface has 2 implementations: StateContextCache and LRUBlockStateCache (new)
- CheckpointStateCache has 2 implementations: CheckpointStateCache and PersistentCheckpointStateCache (new)
- LRUBlockStateCache
  - Bounded by a max number of states, prune on add() to make its size consistent, oldest state is removed first
  - Head state always stay on top to make sure we always have it
- PersistentCheckpointStateCache
  - Only states of the last n epochs are kept in memory, persist remaining states to db
  - Configured by `maxEpochsInMemory`
  - Given a block hex and its state, process that state to decide which checkpoint states are pruned based on its view

This PR only implements new state caches so it'll not affect production code, next PRs will consume them

part of #5968

### CODE REVIEW

File: packages/beacon-node/src/chain/stateCache/lruBlockStateCache.ts [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+  clear(): void {
+    this.cache.clear();
+  }
+
+  get size(): number {
+    return this.cache.size;
+  }
+
+  /**
+   * If a recent state is not available, regen from the checkpoint state.
```
Comment (dapplion) [inline packages/beacon-node/src/chain/stateCache/lruBlockStateCache.ts]: is this a proper description for this function?

File: packages/beacon-node/src/metrics/metrics/lodestar.ts [hunk 2]
```diff
@@ -1074,6 +1074,40 @@ export function createLodestarMetrics(
         help: "Histogram of cloned count per state every time state.clone() is called",
         buckets: [1, 2, 5, 10, 50, 250],
       }),
+      statePersistDuration: register.histogram({
+        name: "lodestar_cp_state_cache_state_persist_seconds",
+        help: "Histogram of time to persist state to memory",
```
Comment (dapplion) [inline packages/beacon-node/src/metrics/metrics/lodestar.ts]: persist **from** memory right? If so, this buckets don't have enough granularity and more bucket values. Same for most bucket selection below too

File: packages/beacon-node/test/unit/chain/stateCache/persistentCheckpointsCache.test.ts [hunk 3]
```diff
@@ -0,0 +1,849 @@
...
+// root0b is of the first slot of epoch 20
+const cp0b = {epoch: 20, root: root0b};
+const cp1 = {epoch: 21, root: root1};
+const cp2 = {epoch: 22, root: root2};
+const [cp0aHex, cp0bHex, cp1Hex, cp2Hex] = [cp0a, cp0b, cp1, cp2].map((cp) => toCheckpointHex(cp));
+const persistent0bKey = toHexString(checkpointToPersistentKey(cp0b));
+const startSlotEpoch20 = computeStartSlotAtEpoch(20);
+const startSlotEpoch21 = computeStartSlotAtEpoch(21);
+const startSlotEpoch22 = computeStartSlotAtEpoch(22);
+const allStates = [cp0a, cp0b, cp1, cp2]
```
Comment (dapplion) [inline packages/beacon-node/test/unit/chain/stateCache/persistentCheckpointsCache.test.ts]: This code is eagerly executed when loading tests, it should only run as part of the tests. If you want a common variable for tests: either memoize it or compute it in a before() block

File: packages/beacon-node/src/chain/stateCache/lruBlockStateCache.ts [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+     +-----+-----+
+     |           |
+  Block 11   Block 12
+     ^           |
+     |           |
+    head       Block 13
+ *
+ * The maintained key order would be: 11 -> 13 -> 12 -> 10, and state 10 will be pruned first.
+ */
+export class LRUBlockStateCache implements BlockStateCache {
```
Comment (dapplion) [inline packages/beacon-node/src/chain/stateCache/lruBlockStateCache.ts]: This strategy is cool! simple enough way to merge the head strong reference with an LRU cache. Where is `LRUBlockStateCache` meant to be used latter?

File: packages/beacon-node/src/chain/stateCache/persistentCheckpointsCache.ts [hunk 5]
```diff
@@ -0,0 +1,573 @@
...
+          if (isPersistentKey(stateOrPersistentKey)) {
+            persistCount++;
+            persistentEpochs.add(epoch);
+          } else {
+            stateCount++;
+            memoryEpochs.add(epoch);
+          }
+        }
+        metrics.cpStateCache.size.set({type: CacheType.persistence}, persistCount);
+        metrics.cpStateCache.size.set({type: CacheType.state}, stateCount);
```
Comment (dapplion) [inline packages/beacon-node/src/chain/stateCache/persistentCheckpointsCache.ts]: I would use better tags: ['persisted', 'in-memory']

File: packages/beacon-node/src/chain/stateCache/persistentCheckpointsCache.ts [hunk 6]
```diff
@@ -0,0 +1,573 @@
...
+
+    if (!isPersistentKey(persistentKey)) {
+      // should not happen, in-memory state is handled above
+      throw new Error("Expected persistent key");
+    }
+
+    const logMeta = {persistentKey: toHexString(persistentKey)};
+
+    // reload from disk or db based on closest checkpoint
+    this.logger.verbose("Reload: read state", logMeta);
```
Comment (dapplion) [inline packages/beacon-node/src/chain/stateCache/persistentCheckpointsCache.ts]: This logs would qualify as debug

File: packages/beacon-node/src/chain/stateCache/persistentCheckpointsCache.ts [hunk 7]
```diff
@@ -0,0 +1,573 @@
...
+    const newStateBytes = await this.persistentApis.read(persistentKey);
+    if (newStateBytes === null) {
+      this.logger.verbose("Reload: read state failed", logMeta);
+      return null;
```
Comment (dapplion) [inline packages/beacon-node/src/chain/stateCache/persistentCheckpointsCache.ts]: Is this an expected state? DB inconsistency should be at least a warn

File: packages/beacon-node/src/chain/stateCache/persistentCheckpointsCache.ts [hunk 8]
```diff
@@ -0,0 +1,573 @@
...
+      // only remove persisted state once we reload successfully
+      this.persistentApis
+        .remove(persistentKey)
+        .catch((e) => this.logger.debug("Error removing persisted state", logMeta, e));
+      this.cache.set(cpKey, newCachedState);
+      // don't prune from memory here, call it at the last 1/3 of slot 0 of an epoch
+      return newCachedState;
+    } catch (e) {
+      this.logger.debug("Error reloading state from disk", logMeta, e as Error);
+      return null;
```
Comment (dapplion) [inline packages/beacon-node/src/chain/stateCache/persistentCheckpointsCache.ts:202]: null here can mean many things at once, why not propagate the error up?

File: packages/beacon-node/src/chain/stateCache/persistentCheckpointsCache.ts [hunk 9]
```diff
@@ -0,0 +1,573 @@
...
+      timer?.();
+      this.logger.verbose("Reload state successfully", {
+        ...logMeta,
+        stateSlot: newCachedState.slot,
+        seedSlot: closestState.slot,
+      });
+      // only remove persisted state once we reload successfully
+      this.persistentApis
+        .remove(persistentKey)
+        .catch((e) => this.logger.debug("Error removing persisted state", logMeta, e));
```
Comment (dapplion) [inline packages/beacon-node/src/chain/stateCache/persistentCheckpointsCache.ts]: Is it necessary to call remove here? Can it be scheduled to be removed in a prune step?

File: packages/beacon-node/src/chain/stateCache/persistentCheckpointsCache.ts [hunk 10]
```diff
@@ -0,0 +1,573 @@
...
+      .filter((e) => e <= maxEpoch);
+    for (const epoch of epochs) {
+      if (this.epochIndex.get(epoch)?.has(rootHex)) {
+        try {
+          const state = await this.getOrReload({rootHex, epoch});
+          if (state) {
+            return state;
+          }
+        } catch (e) {
+          this.logger.debug("Error get or reload state", {epoch, rootHex}, e as Error);
```
Comment (dapplion) [inline packages/beacon-node/src/chain/stateCache/persistentCheckpointsCache.ts:338]: getOrReload already swallows all errors, I need to look more into the usage pattern of this but on a cursory look error handling should be more explicit than converting all errors to null

File: packages/beacon-node/src/chain/stateCache/persistentCheckpointsCache.ts [hunk 11]
```diff
@@ -0,0 +1,573 @@
...
+   *
+   *
+   * - 1 then we'll persist {root: b1, epoch n-1} checkpoint state to disk. Note that at epoch n there is both {root: b0, epoch: n} and {root: c0, epoch: n} checkpoint states in memory
+   * - 2 then we'll persist {root: b2, epoch n-2} checkpoint state to disk, there are also 2 checkpoint states in memory at epoch n, same to the above (maxEpochsInMemory=1)
+   *
+   * As of Nov 2023, it takes 1.3s to 1.5s to persist a state on holesky on fast server. TODO:
+   * - improve state serialization time
+   * - or research how to only store diff against the finalized state
+   */
+  async processState(blockRootHex: RootHex, state: CachedBeaconStateAllForks): Promise<number> {
```
Comment (dapplion) [inline packages/beacon-node/src/chain/stateCache/persistentCheckpointsCache.ts:428]: please add vertical whitespace in this function, it's quite dense and hard to follow?

File: packages/beacon-node/src/chain/stateCache/persistentCheckpointsCache.ts [hunk 12]
```diff
@@ -0,0 +1,573 @@
...
+    let persistCount = 0;
+    // it's important to sort the epochs in ascending order, in case of big reorg we always want to keep the most recent checkpoint states
+    const sortedEpochs = Array.from(this.epochIndex.keys()).sort((a, b) => a - b);
+    while (sortedEpochs.length > this.maxEpochsInMemory) {
```
Comment (dapplion) [inline packages/beacon-node/src/chain/stateCache/persistentCheckpointsCache.ts]: This can be replaced with a for loop. My understanding of the current code:
- read first N items of sortedEpochs starting at index 0, where N = sortedEpochs.length - this.maxEpochsInMemory

This can be emulated by for loop conditions and then you don't need to handle the lowestEpoch === undefined case

File: packages/beacon-node/src/chain/stateCache/persistentCheckpointsCache.ts [hunk 13]
```diff
@@ -0,0 +1,573 @@
...
+  return {
+    rootHex,
+    epoch: Number(epoch),
+  };
+}
+
+function isPersistentKey(
+  stateOrPersistentKey: CachedBeaconStateAllForks | PersistentKey
+): stateOrPersistentKey is PersistentKey {
+  return (stateOrPersistentKey as CachedBeaconStateAllForks).epochCtx === undefined;
```
Comment (dapplion) [inline packages/beacon-node/src/chain/stateCache/persistentCheckpointsCache.ts]: Consider wrapping the items in the cache in an enum style to have a cleaner separation of types:

```ts
type CacheItem = {type: "persistent", value: X} | {type: "in-memory", value: Y}
```

**Review by dapplion — COMMENTED** [review]
Did an initial pass, let's sync @tuyennhv for a deeper review

File: packages/beacon-node/src/chain/stateCache/lruBlockStateCache.ts (hunk 4 above)
Comment (twoeths) [reply packages/beacon-node/src/chain/stateCache/lruBlockStateCache.ts]: this is another/new implementation of `BlockStateCache`, same functionality to `StateContextCache`. This will be used in regen, will have feature flag to decide this new state cache or the old `StateContextCache` in the next PR

File: packages/beacon-node/src/chain/stateCache/lruBlockStateCache.ts (hunk 1 above)
Comment (twoeths) [reply packages/beacon-node/src/chain/stateCache/lruBlockStateCache.ts]: yeah this is not a good description, will revise it 👍

File: packages/beacon-node/src/chain/stateCache/persistentCheckpointsCache.ts (hunk 9 above)
Comment (twoeths) [reply packages/beacon-node/src/chain/stateCache/persistentCheckpointsCache.ts]: yes, in that case we need to model the in-memory cache item as:
```typescript
type InMemoryCacheItem = {
  type: CacheType.inMemory;
  state: CachedBeaconStateAllForks;
  // this is set when we reload a state
  persistedKey?: PersistedKey
};
```

so that when finalized checkpoint, we prune based on the above reloaded `persistedKey`

File: packages/beacon-node/src/chain/stateCache/persistentCheckpointsCache.ts (hunk 10 above)
Comment (twoeths) [reply packages/beacon-node/src/chain/stateCache/persistentCheckpointsCache.ts:338]: in `regen.get*()` apis, they only expect checkpoint state caches returning state or `null`. We could shift the error handling there but I think regen should also have to try/catch and try to load another checkpoint state. It depends on how we decide the api spec in not only this implementation but the generalized `CheckpointStateCache` interface and all of the implementations so that we have no issue switching the old `CheckpointStateCache` vs this new `PersistentCheckpointStateCache`

File: packages/beacon-node/src/chain/stateCache/types.ts [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+};
+
+export type PersistentCheckpointStateCacheModules = {
+  metrics?: Metrics | null;
+  logger: Logger;
+  clock?: IClock | null;
+  shufflingCache: ShufflingCache;
+  persistentApis: CPStatePersistentApis;
+  getHeadState?: GetHeadStateFn;
+};
```
Comment (wemeetagain) [inline packages/beacon-node/src/chain/stateCache/types.ts]: Might be better to move these to persistentCheckpointsCache.ts since they're not really shared.

File: packages/beacon-node/src/chain/stateCache/lruBlockStateCache.ts [hunk 15]
```diff
@@ -0,0 +1,174 @@
+import {toHexString} from "@chainsafe/ssz";
```
Comment (wemeetagain) [inline packages/beacon-node/src/chain/stateCache/lruBlockStateCache.ts]: this is very nice, but I think the name LRU is misleading.
IMO this is a FIFO cache with special handling for the head.
I'd rename instances of LRU (and lru) to FIFO (and fifo).

File: packages/beacon-node/src/chain/stateCache/persistentCheckpointsCache.ts [hunk 16]
```diff
@@ -0,0 +1,607 @@
...
+    const logMeta = {persistedKey: toHexString(persistedKey)};
+
+    // reload from disk or db based on closest checkpoint
+    this.logger.debug("Reload: read state", logMeta);
+    const newStateBytes = await this.persistentApis.read(persistedKey);
+    if (newStateBytes === null) {
+      this.logger.warn("Reload: read state failed", logMeta);
+      return null;
+    }
+    this.logger.debug("Reload: read state successfully", logMeta);
```
Comment (wemeetagain) [inline packages/beacon-node/src/chain/stateCache/persistentCheckpointsCache.ts]:
```suggestion
    this.logger.debug("Reload: read state successful", logMeta);
```

File: packages/beacon-node/src/chain/stateCache/persistentCheckpointsCache.ts [hunk 17]
```diff
@@ -0,0 +1,607 @@
...
+      // should not happen, in-memory state is handled above
+      throw new Error("Expected persisted key");
+    }
+
+    const persistedKey = cacheItem.value;
+    const logMeta = {persistedKey: toHexString(persistedKey)};
+
+    // reload from disk or db based on closest checkpoint
+    this.logger.debug("Reload: read state", logMeta);
+    const newStateBytes = await this.persistentApis.read(persistedKey);
```
Comment (wemeetagain) [inline packages/beacon-node/src/chain/stateCache/persistentCheckpointsCache.ts]: is it worth timing this?

File: packages/beacon-node/src/chain/stateCache/persistentCheckpointsCache.ts [hunk 18]
```diff
@@ -0,0 +1,607 @@
...
+      return null;
+    }
+    this.logger.debug("Reload: read state successfully", logMeta);
+    this.metrics?.stateReloadSecFromSlot.observe(this.clock?.secFromSlot(this.clock?.currentSlot ?? 0) ?? 0);
+    const closestState = this.findSeedStateToReload(cp) ?? this.getHeadState?.();
+    if (closestState == null) {
+      throw new Error("No closest state found for cp " + toCheckpointKey(cp));
+    }
+    this.metrics?.stateReloadEpochDiff.observe(Math.abs(closestState.epochCtx.epoch - cp.epoch));
+    this.logger.debug("Reload: found closest state", {...logMeta, seedSlot: closestState.slot});
```
Comment (wemeetagain) [inline packages/beacon-node/src/chain/stateCache/persistentCheckpointsCache.ts]:
```suggestion
    this.logger.debug("Reload: found closest seed state", {...logMeta, seedSlot: closestState.slot});
```

File: packages/beacon-node/src/chain/stateCache/persistentCheckpointsCache.ts [hunk 19]
```diff
@@ -0,0 +1,607 @@
...
+    const timer = this.metrics?.stateReloadDuration.startTimer();
+
+    try {
+      const newCachedState = loadCachedBeaconState(closestState, newStateBytes, {
+        shufflingGetter: this.shufflingCache.getSync.bind(this.shufflingCache),
+      });
+      timer?.();
+      this.logger.debug("Reload state successfully", {
```
Comment (wemeetagain) [inline packages/beacon-node/src/chain/stateCache/persistentCheckpointsCache.ts]:
```suggestion
      this.logger.debug("Reload: cached state load successful", {
```

File: packages/beacon-node/src/chain/stateCache/persistentCheckpointsCache.ts [hunk 20]
```diff
@@ -0,0 +1,607 @@
...
+        ...logMeta,
+        stateSlot: newCachedState.slot,
+        seedSlot: closestState.slot,
+      });
+      // only remove persisted state once we reload successfully
+      this.cache.set(cpKey, {type: CacheType.inMemory, state: newCachedState, persistedKey});
+      // don't prune from memory here, call it at the last 1/3 of slot 0 of an epoch
+      return newCachedState;
+    } catch (e) {
+      this.logger.debug("Error reloading state from disk", logMeta, e as Error);
```
Comment (wemeetagain) [inline packages/beacon-node/src/chain/stateCache/persistentCheckpointsCache.ts]:
```suggestion
      this.logger.debug("Reload: error loading cached state", logMeta, e as Error);
```

File: packages/beacon-node/src/chain/stateCache/persistentCheckpointsCache.ts [hunk 21]
```diff
@@ -0,0 +1,607 @@
...
+  }
+
+  /**
+   * Get a state from cache, it may reload from disk.
+   * This is an expensive api, should only be called in some important flows:
+   * - Validate a gossip block
+   * - Get block for processing
+   * - Regen head state
+   */
+  async getOrReload(cp: CheckpointHex): Promise<CachedBeaconStateAllForks | null> {
```
Comment (wemeetagain) [inline packages/beacon-node/src/chain/stateCache/persistentCheckpointsCache.ts:163]: can this method reuse `getStateOrBytes`?

File: packages/beacon-node/src/chain/stateCache/persistentCheckpointsCache.ts (hunk 17 above)
Comment (twoeths) [reply packages/beacon-node/src/chain/stateCache/persistentCheckpointsCache.ts]: yes will also track it 👍

File: packages/beacon-node/src/chain/stateCache/persistentCheckpointsCache.ts (hunk 21 above)
Comment (twoeths) [reply packages/beacon-node/src/chain/stateCache/persistentCheckpointsCache.ts:163]: I refactor to a new method `getStateOrLoadDb()` to be used by `getOrReload()` and `getStateOrBytes` please have a look

**Review by wemeetagain — APPROVED** [review]
lgtm

**Comment by wemeetagain** [comment]
:tada: This PR is included in [v1.14.0](https://github.com/ChainSafe/lodestar/releases/tag/v1.14.0) :tada:
